### PR TITLE
Optional inconsistency in Naive Bayes #1565

### PR DESCRIPTION
**Pull request by whoopityDoop**

### CODE REVIEW

File: src/mlpack/methods/kmeans/kmeans_main.cpp [hunk 1]
```diff
@@ -28,17 +28,11 @@ using namespace mlpack::util;
...
-    "An implementation of several strategies for efficient k-means clustering. "
-    "Given a dataset and a value of k, this computes and returns a k-means "
-    "clustering on that data.",
-    // Long description.
-    "This program performs K-Means clustering on the given dataset.  It can "
-    "return the learned cluster assignments, and the centroids of the clusters."
-    "  Empty clusters are not allowed by default; when a cluster becomes empty,"
-    " the point furthest from the centroid of the cluster with maximum variance"
-    " is taken to fill that cluster."
+PROGRAM_INFO("K-Means Clustering", "This program performs K-Means clustering "
```
[the remark this answers is not in the document]
Comment (shashank-007) [reply src/mlpack/methods/kmeans/kmeans_main.cpp]: @zoq The version of mlpack i am working on my computer was not having this short long description.

[the remark this answers is not in the document]
Comment (zoq) [reply src/mlpack/methods/kmeans/kmeans_main.cpp]: I see, can you update the PR against the latest version?

[the remark this answers is not in the document]
Comment (shashank-007) [reply src/mlpack/methods/kmeans/kmeans_main.cpp]: @zoq  while i am working with the latest version why  would this be causing macro errors 
error: macro "PROGRAM_INFO" passed 10 arguments, but takes just 2
         "@doxygen/classmlpack_1_1kmeans_1_1KMeans.html"));
                                                         ^
/home/whoopitydoop/mlpack-3.0.4/src/mlpack/methods/kmeans/kmeans_main.cpp:31:1: error: ‘PROGRAM_INFO’ does not name a type
 PROGRAM_INFO("K-Means Clustering",
 ^~~~~~~~~~~~

[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/methods/kmeans/kmeans_main.cpp]: Don't copy paste in the changes to your branch; instead merge the master branch into your branch.  Or make a new branch against current master and put your changes there.

File: src/mlpack/methods/kmeans/kmeans_main.cpp [hunk 2]
```diff
@@ -91,21 +91,7 @@ PROGRAM_INFO("K-Means Clustering",
...
-        "http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.586.2554"
-        "&rep=rep1&type=pdf"),
-    SEE_ALSO("Accelerating exact k-means algorithms with geometric reasoning "
-        "(pdf)", "http://reports-archive.adm.cs.cmu.edu/anon/anon/usr/ftp/"
-        "usr0/ftp/2000/CMU-CS-00-105.pdf"),
-    SEE_ALSO("A dual-tree algorithm for fast k-means clustering with large k "
-        "(pdf)", "http://www.ratml.org/pub/pdf/2017dual.pdf"),
-    SEE_ALSO("mlpack::kmeans::KMeans class documentation",
-        "@doxygen/classmlpack_1_1kmeans_1_1KMeans.html"));
+        "clusters", 10, "max_iterations", 500, "centroid", "final"));
```
Comment (zoq) [inline src/mlpack/methods/kmeans/kmeans_main.cpp]: Looks like the fork you are working on is still outdated. One option would be to update your fork, see https://stackoverflow.com/questions/7244321/how-do-i-update-a-github-forked-repository for more information about how to update a fork.

Comment (shashank-007) [reply src/mlpack/methods/kmeans/kmeans_main.cpp]: @zoq i updated the fork .now make mlpack_test and individual tests are running i checked .

**Comment by shashank-007** [comment]
@rcurtin @zoq any reason that Travis cli test is failing i tried to look into but couldn't find one. Any directions on where to look for

File: src/mlpack/methods/kmeans/kmeans_main.cpp (hunk 2 above)
Comment (rcurtin) [reply src/mlpack/methods/kmeans/kmeans_main.cpp]: The changes should not remove the `SEE_ALSO()`s, can you fix this please?

**Comment by rcurtin** [comment]
@shashank-007: what have you tried?  It gives output:

```
[FATAL] Parameter --output does not exist in this program!
unknown location(0): fatal error in "NBCOutputDimensionTest": std::runtime_error: fatal error; see Log::Fatal output
/home/travis/build/mlpack/mlpack/src/mlpack/tests/main_tests/nbc_test.cpp(83): last checkpoint
```

Have you tried looking in that test file?

Also, we need to preserve the old options for reverse compatibility.  Maybe it would be a good idea to take a look at #1616 for some ideas.

**Comment by shashank-007** [comment]
@rcurtin i looked into the file there is no such variable named as output or a function which references it .perhaps i am missing something

**Comment by rcurtin** [comment]
Try using a debugger to step through the code in order to find where the exception is being thrown.

**Comment by shashank-007** [comment]
@rcurtin i should use debugger on 
```
bin/mlpack_test ??
```
If Yes,then how ??

**Comment by shashank-007** [comment]
@rcurtin i fixed the compatibility issue you mentioned along with the error..

File: src/mlpack/methods/kmeans/kmeans_main.cpp [hunk 3]
```diff
@@ -91,21 +91,7 @@ PROGRAM_INFO("K-Means Clustering",
     "following command may be used:"
     "\n\n" +
     PRINT_CALL("kmeans", "input", "data", "initial_centroids", "initial",
-        "clusters", 10, "max_iterations", 500, "centroid", "final"),
-    SEE_ALSO("K-Means tutorial", "@doxygen/kmtutorial.html"),
```
Comment (zoq) [inline src/mlpack/methods/kmeans/kmeans_main.cpp]: Hm, this changes are still missing, did you rebased the PR against the latest master branch?

File: src/mlpack/methods/naive_bayes/nbc_main.cpp [hunk 4]
```diff
@@ -188,14 +191,17 @@ static void mlpackMain()
     model->nbc.Classify(testingData, predictions, probabilities);
     Timer::Stop("nbc_testing");
 
-    if (CLI::HasParam("output"))
+    if (CLI::HasParam("output")||CLI::HasParam("predictions"))
```
Comment (zoq) [inline src/mlpack/methods/naive_bayes/nbc_main.cpp]: Missing space between ``||`` on both sides.

File: src/mlpack/methods/naive_bayes/nbc_main.cpp [hunk 5]
```diff
@@ -123,9 +125,10 @@ static void mlpackMain()
   RequireOnlyOnePassed({ "training", "input_model" }, true);
   ReportIgnoredParam({{ "training", false }}, "labels");
   ReportIgnoredParam({{ "training", false }}, "incremental_variance");
-  RequireAtLeastOnePassed({ "output", "output_model", "output_probs" }, false,
+  RequireAtLeastOnePassed({ "output","predictions","output_model", "output_probs" }, false,
```
Comment (zoq) [inline src/mlpack/methods/naive_bayes/nbc_main.cpp]: Missing space right after the comma.

File: src/mlpack/methods/kmeans/kmeans_main.cpp (hunk 3 above)
Comment (whoopityDoop) [reply src/mlpack/methods/kmeans/kmeans_main.cpp]: i did rebased the pr against the latest master branch.Playing around with the code i may have replaced the latest version .

File: src/mlpack/methods/naive_bayes/nbc_main.cpp (hunk 5 above)
Comment (whoopityDoop) [reply src/mlpack/methods/naive_bayes/nbc_main.cpp]: Corrected please take a look at it again .

File: src/mlpack/methods/naive_bayes/nbc_main.cpp (hunk 4 above)
Comment (whoopityDoop) [reply src/mlpack/methods/naive_bayes/nbc_main.cpp]: Done long with some more issues encountered

File: src/mlpack/core/tree/spill_tree/spill_tree_impl.hpp [hunk 6]
```diff
@@ -674,8 +674,8 @@ bool SpillTree<MetricType, StatisticType, MatType, HyperplaneType, SplitType>::
   const double p1 = (double) (left + rightFrontier) / points.n_elem;
   const double p2 = (double) (right + leftFrontier) / points.n_elem;
 
-  if ((p1 <= rho || rightFrontier == 0) &&
-      (p2 <= rho || leftFrontier == 0))
+  if ((p1 <= this->NumChildren()*rho || rightFrontier == 0) &&
+      (p2 <= this->NumChildren()*rho || leftFrontier == 0) && tau != 0)
```
Comment (rcurtin) [inline src/mlpack/core/tree/spill_tree/spill_tree_impl.hpp]: Are you sure this is supposed to be a part of this PR?

File: src/mlpack/methods/kmeans/kmeans_main.cpp [hunk 7]
```diff
@@ -131,43 +128,36 @@ PARAM_INT_IN("max_iterations", "Maximum number of iterations before k-means "
...
 // the Lloyd iteration step type and run k-means.
 template<typename InitialPartitionPolicy, typename EmptyClusterPolicy>
 void FindLloydStepType(const InitialPartitionPolicy& ipp);
-
 // Given the template parameters, sanitize/load input and run k-means.
 template<typename InitialPartitionPolicy,
          typename EmptyClusterPolicy,
          template<class, class> class LloydStepType>
 void RunKMeans(const InitialPartitionPolicy& ipp);
-
```
Comment (rcurtin) [inline src/mlpack/methods/kmeans/kmeans_main.cpp]: Actually all these spacing lines were here to make the code split into logical blocks.  Can we leave the blank lines there please?

File: src/mlpack/methods/naive_bayes/nbc_main.cpp [hunk 8]
```diff
@@ -112,8 +111,11 @@ PARAM_FLAG("incremental_variance", "The variance of each class will be "
 
 // Test parameters.
 PARAM_MATRIX_IN("test", "A matrix containing the test set.", "T");
+// parameter output is to be deprecated and removed in mlpack 4.0
 PARAM_UROW_OUT("output", "The matrix in which the predicted labels for the"
     " test set will be written.", "o");
```
Comment (rcurtin) [inline src/mlpack/methods/naive_bayes/nbc_main.cpp]: We should point out that `output` is deprecated.

File: src/mlpack/methods/naive_bayes/nbc_main.cpp [hunk 9]
```diff
@@ -112,8 +111,11 @@ PARAM_FLAG("incremental_variance", "The variance of each class will be "
...
+PARAM_UROW_OUT("predictions", "The matrix in which the predicted labels for the"
+    " test set will be written.", "a");
 PARAM_MATRIX_OUT("output_probs", "The matrix in which the predicted probability"
     " of labels for the test set will be written.", "p");
```
Comment (rcurtin) [inline src/mlpack/methods/naive_bayes/nbc_main.cpp]: We should also deprecate `output_probs` and provide a `probabilities` output parameter.

File: src/mlpack/methods/naive_bayes/nbc_main.cpp [hunk 10]
```diff
@@ -188,14 +191,17 @@ static void mlpackMain()
...
       Row<size_t> rawResults;
       data::RevertLabels(predictions, model->mappings, rawResults);
 
-      // Output results.
-      CLI::GetParam<Row<size_t>>("output") = std::move(rawResults);
+      CLI::GetParam<Row<size_t>>("predictions") = std::move(rawResults);
+      if (CLI::HasParam("output"))
+      {
+         CLI::GetParam<Row<size_t>>("output") = std::move(predictions);
+      }
```
Comment (rcurtin) [inline src/mlpack/methods/naive_bayes/nbc_main.cpp]: Why are you setting one to `predictions` and one to `rawResults`?  I think both should be set to `rawResults`.

File: src/mlpack/methods/naive_bayes/nbc_main.cpp [hunk 11]
```diff
@@ -123,9 +125,10 @@ static void mlpackMain()
   RequireOnlyOnePassed({ "training", "input_model" }, true);
   ReportIgnoredParam({{ "training", false }}, "labels");
   ReportIgnoredParam({{ "training", false }}, "incremental_variance");
-  RequireAtLeastOnePassed({ "output", "output_model", "output_probs" }, false,
-      "no output will be saved");
+  RequireAtLeastOnePassed({ "output", "predictions", "output_model",
+   "output_probs" }, false, "no output will be saved");
```
Comment (rcurtin) [inline src/mlpack/methods/naive_bayes/nbc_main.cpp]: This new line should be indented twice. :+1:

File: src/mlpack/methods/naive_bayes/nbc_main.cpp [hunk 12]
```diff
@@ -51,9 +51,9 @@ PROGRAM_INFO("Parametric Naive Bayes Classifier",
...
     "If classifying a test set is desired, the test set may be specified with "
-    "the " + PRINT_PARAM_STRING("test") + " parameter, and the "
-    "classifications may be saved with the " + PRINT_PARAM_STRING("output") +
-    " output parameter.  If saving the trained model is desired, this may be "
+    "the " + PRINT_PARAM_STRING("test") + " parameter, and the classifications"
+    " may be saved with the " + PRINT_PARAM_STRING("predictions") +"predictions"
+    "  parameter.  If saving the trained model is desired, this may be "
     "done with the " + PRINT_PARAM_STRING("output_model") + " output "
     "parameter."
     "\n\n"
```
Comment (rcurtin) [inline src/mlpack/methods/naive_bayes/nbc_main.cpp]: Can we add a paragraph pointing out that the `output` (and `output_probs` when you do it) options are deprecated and will be removed in mlpack 4.0.0?  You can take a look at `logistic_regression_main.cpp` to get an idea of what I mean.

File: src/mlpack/methods/naive_bayes/nbc_main.cpp [hunk 13]
```diff
@@ -79,7 +79,6 @@ PROGRAM_INFO("Parametric Naive Bayes Classifier",
     SEE_ALSO("mlpack::naive_bayes::NaiveBayesClassifier C++ class "
         "documentation", "@doxygen/classmlpack_1_1naive__bayes_1_1"
         "NaiveBayesClassifier.html"));
-
```
Comment (rcurtin) [inline src/mlpack/methods/naive_bayes/nbc_main.cpp]: We should leave this line here too.

File: src/mlpack/methods/naive_bayes/nbc_main.cpp [hunk 14]
```diff
@@ -112,8 +111,11 @@ PARAM_FLAG("incremental_variance", "The variance of each class will be "
 
 // Test parameters.
 PARAM_MATRIX_IN("test", "A matrix containing the test set.", "T");
+// parameter output is to be deprecated and removed in mlpack 4.0
```
Comment (rcurtin) [inline src/mlpack/methods/naive_bayes/nbc_main.cpp]: I know this is picky, but can we be sure to capitalize and punctuate the sentence here so it matches mlpack commenting guidelines?  Like this:

```
// The parameter 'output' is deprecated and will be removed in mlpack 4.0.0.
```

Thanks! :+1:

File: src/mlpack/core/tree/spill_tree/spill_tree_impl.hpp (hunk 6 above)
Comment (whoopityDoop) [reply src/mlpack/core/tree/spill_tree/spill_tree_impl.hpp]: No,i was working on spill trees issues .i forgot to remove it .

File: src/mlpack/methods/naive_bayes/nbc_main.cpp (hunk 10 above)
Comment (whoopityDoop) [reply src/mlpack/methods/naive_bayes/nbc_main.cpp]: On running the tests if we set both to rawResults tests would fail whereas this way they will pass and algorithm will run correctly too

Comment (whoopityDoop) [reply src/mlpack/methods/naive_bayes/nbc_main.cpp]: @rcurtin 
Perhaps,i could do this too
` ` `
predictions=rawResults;
if(CLI::HasParam("output"))
{
    CLI::GetParam<Row<size_t>>("output") = std::move(predictions);
 }
` ` `
this will work too

File: src/mlpack/methods/naive_bayes/nbc_main.cpp (hunk 12 above)
Comment (whoopityDoop) [reply src/mlpack/methods/naive_bayes/nbc_main.cpp]: @rcurtin i have a question so wouldn't the changes i made be visible while i use mlpack_nbc --h ?

**Comment by whoopityDoop** [comment]
@rcurtin tests are completed

File: src/mlpack/methods/naive_bayes/nbc_main.cpp (hunk 12 above)
Comment (rcurtin) [reply src/mlpack/methods/naive_bayes/nbc_main.cpp]: Yes, when you recompile, the changes you make here should be visible when you run `mlpack_nbc -h`.

File: src/mlpack/methods/naive_bayes/nbc_main.cpp [hunk 15]
```diff
@@ -112,20 +119,27 @@ PARAM_FLAG("incremental_variance", "The variance of each class will be "
...
 static void mlpackMain()
 {
   // Check input parameters.
   RequireOnlyOnePassed({ "training", "input_model" }, true);
   ReportIgnoredParam({{ "training", false }}, "labels");
   ReportIgnoredParam({{ "training", false }}, "incremental_variance");
-  RequireAtLeastOnePassed({ "output", "output_model", "output_probs" }, false,
-      "no output will be saved");
+  RequireAtLeastOnePassed({ "output", "predictions", "output_model",
+     "output_probs", "probabilities" }, false, "no output will be saved");
```
Comment (rcurtin) [inline src/mlpack/methods/naive_bayes/nbc_main.cpp]: There's a little style issue here... a wrapped line should be indented twice (four spaces) but this looks like 3 spaces.

File: src/mlpack/methods/naive_bayes/nbc_main.cpp (hunk 13 above)
Comment (rcurtin) [reply src/mlpack/methods/naive_bayes/nbc_main.cpp]: Can you re-add the blank line please?  Vertical whitespace is important to logically break up the code and make it easier to read.

File: src/mlpack/methods/naive_bayes/nbc_main.cpp (hunk 10 above)
Comment (rcurtin) [reply src/mlpack/methods/naive_bayes/nbc_main.cpp]: The algorithm does not run correctly if you set one to `rawResults` and one to `predictions` because those both contain different values.

The updated code still has a problem.  The `if` statement outside this block assumes that either `CLI::HasParam("output")` is true or `CLI::HasParam("predictions")` is true.  But then inside, it assumes that `predictions` has been specified and calls `GetParam()` on it (even if it has not been specified).  Can you clean this up?

Also, the line `predictions = rawResults` makes the entire code wrong.  Did you read the documentation for `data::RevertLabels()` to understand what it does and why that line is there?  If you do, you will understand why `predictions` and `rawResults` are different and should be treated as such.

File: src/mlpack/methods/naive_bayes/nbc_main.cpp [hunk 16]
```diff
@@ -188,17 +201,23 @@ static void mlpackMain()
...
+      CLI::GetParam<Row<size_t>>("predictions") = predictions;
+      if (CLI::HasParam("output"))
+      {
+        CLI::GetParam<Row<size_t>>("output") = std::move(predictions);
+      }
+    }
+    CLI::GetParam<mat>("probabilities") = probabilities;
+    if (CLI::HasParam("output_probs"))
+    {
+      CLI::GetParam<mat>("output_probs") = probabilities;
```
Comment (rcurtin) [inline src/mlpack/methods/naive_bayes/nbc_main.cpp]: This code does not check if `CLI::HasParam("probabilities")` is true, and when you use `probabilities` the second time you can use `std::move()`.  Please read the documentation for `std::move()` (and move semantics) to understand why. :+1:

File: src/mlpack/tests/main_tests/nbc_test.cpp [hunk 17]
```diff
@@ -331,4 +331,110 @@ BOOST_AUTO_TEST_CASE(NBCIncrementalVarianceTest)
...
 
+/**
+  * Ensuring that the parameter 'output' and the parameter 'predictions' give
+  * the same output.
+ **/
+// The following test case is to check whether the old parameter 'output' and
+// the new parameter 'predictions' give the same output.
+// This test case will be removed in mlpack4
+// when the deprecated parameter: 'output' is removed.
+BOOST_AUTO_TEST_CASE(NBCOPtionConsistencyTest)
```
Comment (rcurtin) [inline src/mlpack/tests/main_tests/nbc_test.cpp]: Typo: `OPtion` -> `Option`.

File: src/mlpack/methods/naive_bayes/nbc_main.cpp (hunk 16 above)
Comment (whoopityDoop) [reply src/mlpack/methods/naive_bayes/nbc_main.cpp]: Yea,that was a mistake on my part.

**Comment by whoopityDoop** [comment]
@rcurtin i have fixed the issues you mentioned .take a further look

File: src/mlpack/methods/kmeans/kmeans_main.cpp [hunk 18]
```diff
@@ -167,7 +167,6 @@ static void mlpackMain()
     math::RandomSeed((size_t) CLI::GetParam<int>("seed"));
   else
     math::RandomSeed((size_t) std::time(NULL));
-
```
Comment (rcurtin) [inline src/mlpack/methods/kmeans/kmeans_main.cpp]: I don't see a reason to remove this line; can we revert this change please?

File: src/mlpack/methods/naive_bayes/nbc_main.cpp [hunk 19]
```diff
@@ -112,20 +120,27 @@ PARAM_FLAG("incremental_variance", "The variance of each class will be "
...
 PARAM_UROW_OUT("output", "The matrix in which the predicted labels for the"
     " test set will be written.", "o");
+PARAM_UROW_OUT("predictions", "The matrix in which the predicted labels for the"
+    " test set will be written.", "a");
+PARAM_MATRIX_OUT("probabilities", "The matrix in which the predicted"
+    " probability of labels for the test set will be written.", "p");
+// The parameter 'output_probs' is deprecated and can be removed in mlpack 4.0.
 PARAM_MATRIX_OUT("output_probs", "The matrix in which the predicted probability"
-    " of labels for the test set will be written.", "p");
+    " of labels for the test set will be written.", "s");
```
Comment (rcurtin) [inline src/mlpack/methods/naive_bayes/nbc_main.cpp]: I'd be happy to have no alias for the `output_probs` parameter too (i.e. `""` instead of `"s"`).

File: src/mlpack/methods/naive_bayes/nbc_main.cpp (hunk 10 above)
Comment (rcurtin) [reply src/mlpack/methods/naive_bayes/nbc_main.cpp]: > The updated code still has a problem. The if statement outside this block assumes that either CLI::HasParam("output") is true or CLI::HasParam("predictions") is true. But then inside, it assumes that predictions has been specified and calls GetParam() on it (even if it has not been specified). Can you clean this up?

This problem still exists in the code, can you fix it please?

File: src/mlpack/tests/main_tests/nbc_test.cpp [hunk 20]
```diff
@@ -331,4 +331,110 @@ BOOST_AUTO_TEST_CASE(NBCIncrementalVarianceTest)
...
+/**
+  * Ensuring that the parameter 'output_probabilities' and the parameter
+  * 'probabilities' give the same output
+ **/
+// The following test case is to check whether the old parameter
+// 'output_probabilities' and the new parameter 'probabilities' give the same
+// output. This test case will be removed in mlpack 4
+// when the deprecated parameter: 'output_probabilities' is removed
+BOOST_AUTO_TEST_CASE(NBCOptionConsistencyTest2)
+{   arma::mat inputData;
```
Comment (rcurtin) [inline src/mlpack/tests/main_tests/nbc_test.cpp]: Why is a declaration on the same line as an opening brace?  Can you put them on two different lines please?

**Comment by whoopityDoop** [comment]
@rcurtin travisCLi build fails i cannot understand the issue

**Comment by rcurtin** [comment]
Let me restart it and see what happens.  Looks like some timeout issue unrelated to the code...

**Comment by whoopityDoop** [comment]
@rcurtin take a look at the changes now .

File: src/mlpack/tests/main_tests/nbc_test.cpp [hunk 21]
```diff
@@ -331,4 +331,111 @@ BOOST_AUTO_TEST_CASE(NBCIncrementalVarianceTest)
...
 }
 
+/**
+  * Ensuring that the parameter 'output' and the parameter 'predictions' give
+  * the same output.
+ **/
+// The following test case is to check whether the old parameter 'output' and
+// the new parameter 'predictions' give the same output.
+// This test case will be removed in mlpack4
+// when the deprecated parameter: 'output' is removed.
```
Comment (rcurtin) [inline src/mlpack/tests/main_tests/nbc_test.cpp]: Some of this comment is redundant and you can put it all in the block up above, instead of having some as `/* */` and some as `//`.

File: src/mlpack/tests/main_tests/nbc_test.cpp [hunk 22]
```diff
@@ -331,4 +331,111 @@ BOOST_AUTO_TEST_CASE(NBCIncrementalVarianceTest)
...
+
+
+/**
+  * Ensuring that the parameter 'output_probabilities' and the parameter
+  * 'probabilities' give the same output
+ **/
+// The following test case is to check whether the old parameter
+// 'output_probabilities' and the new parameter 'probabilities' give the same
+// output. This test case will be removed in mlpack 4
+// when the deprecated parameter: 'output_probabilities' is removed
```
Comment (rcurtin) [inline src/mlpack/tests/main_tests/nbc_test.cpp]: Same here.

**Comment by rcurtin** [comment]
Thanks for the contribution! :+1:
